rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,223 @@
+- It has a test that checks for connectivity to any read/write dependencies:
+
+  - Databases e.g. MongoDB, Postgres, MySQL, Redis
+  - Queueing systems e.g. RabbitMQ
+  - File systems e.g. AWS S3
+  - Caching systems e.g. Memcached
+
+> Example: [Travel Advice Publisher](https://github.com/alphagov/travel-advice-publisher) has enough of these tests because it has one to check login with Signon [[1](https://github.com/alphagov/smokey/blob/a60df2cde5c886fa6d54e8cd820d90facca89e8e/features/publishing_tools.feature#L106)] and another to check its `/healthcheck` endpoint for connectivity to Redis and MongoDB [[1](https://github.com/alphagov/smokey/blob/a60df2cde5c886fa6d54e8cd820d90facca89e8e/features/travel_advice_publisher.feature)] [[2](https://github.com/alphagov/travel-advice-publisher/pull/979)].
+
+> **Services are special.** We cannot make a request to a service, as it has no web interface. Instead, we will need infer the service is running indirectly, using metrics sent to Graphite in response to [a regular heartbeat](https://github.com/alphagov/publishing-api/blob/a4a054b768b7c98fcd1448f74fe6e975ab69bb2f/config/schedule.rb#L10).
+
+### Security concerns
+
+We need to ensure that only people with [production access](https://docs.publishing.service.gov.uk/manual/rules-for-getting-production-access.html) are able to deploy code to the live site. This is equivalent to being a member of the "GOV.UK Production" GitHub team. As part of the trial of CD, we restricted merge access on the affected repos to people in this team [[1](https://github.com/alphagov/govuk-saas-config/pull/49)] [[2](https://github.com/alphagov/govuk-saas-config/pull/51)] [[3](https://github.com/alphagov/govuk-saas-config/pull/55)]. We will [continue to use this approach](https://github.com/alphagov/govuk-saas-config/pull/52#issuecomment-651903746).
+
+As well as restricting who can merge, we also had to address a vulnerability with Integration Jenkins: someone without production access could manually trigger the CD pipeline by running the [`Deploy_App_Downstream` Jenkins job](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb). In order to prevent this kind of unauthorised deployment, the job has additional checks:
+
+- The tag to deploy must be the highest `release_123`-type tag.
+- The tagged commit must match the latest commit on the master branch.
+
+Note that someone without production access could still subvert this protection, by manually altering the job definition. In this scenario, we would permit an unauthorised deployment to the Staging environment, but no further: only people with production access can access Staging Jenkins [[1](https://github.com/alphagov/govuk-developer-docs/blame/736a167bf93b7338347c73cacbddfbdc68e72733/source/manual/rules-for-getting-production-access.html.md#L18-L19)], which also does the above checks.
```

Comment:
Yes, that's right. It's a compromise, really, since that version of the pipeline would be more complex e.g.

1. "Pre_Deploy_App" job to do the check, calls "Deploy_App".
2. "Deploy_App" is unchanged (to allow for branch deploys, etc.).
3. "Deploy_App_Downstream" calls "Pre_Deploy_App" in the next environment.

While it's not ideal, I do think the current pipeline strikes a good balance between security and simplicity, since the worst case scenario is (a) highly unlikely; and (b) low impact (it's not production).

I'd like to suggest we leave it as-is, in the knowledge that our replatforming will do away with all the funky Jenkins jobs in general - this is perhaps another reason to keep them simple.

Would that be OK?

Reply:
Yeah, seems a reasonable trade-off given current climate and inform replatforming we'd hope the new infrastructure can resolve this risk (and, at the least, not make it riskier without consideration)

Reply:
Sure, I'll resolve this now.

cc @bilbof @sengi, although I'm sure you're already aware of this.